src/composer.ts

```diff
@@ -232,33 +335,108 @@ export class AtomicTransactionComposer {
         );
       }
 
-      appArgTypes.push(argSpec.type);
-      appArgValues.push(argValue);
+      if (abiTypeIsReference(argType)) {
+        refArgIndexToBasicArgIndex.set(
+          refArgTypes.length,
+          basicArgTypes.length
+        );
+        refArgTypes.push(argType);
+        refArgValues.push(argValue);
+        // treat the reference as a uint8 for encoding purposes
+        argType = new ABIUintType(8);
+      }
+
+      if (typeof argType === 'string') {
+        throw new Error(`Unknown ABI type: ${argType}`);
+      }
+
+      basicArgTypes.push(argType);
+      basicArgValues.push(argValue);
+    }
+
+    const resolvedRefIndexes: number[] = [];
+    const foreignAccounts: string[] = [];
+    const foreignApps: number[] = [];
+    const foreignAssets: number[] = [];
+    for (let i = 0; i < refArgTypes.length; i++) {
+      const refType = refArgTypes[i];
+      const refValue = refArgValues[i];
+      let resolved = 0;
+
+      switch (refType) {
+        case ABIReferenceType.account: {
+          const addressType = new ABIAddressType();
+          const address = addressType.decode(addressType.encode(refValue));
+          resolved = populateForeignArray(address, foreignAccounts, sender);
+          break;
+        }
+        case ABIReferenceType.application: {
+          const uint64Type = new ABIUintType(64);
+          const refAppID = uint64Type.decode(uint64Type.encode(refValue));
+          if (refAppID > Number.MAX_SAFE_INTEGER) {
+            throw new Error(
+              `Expected safe integer for application value, got ${refAppID}`
+            );
+          }
+          resolved = populateForeignArray(Number(refAppID), foreignApps, appID);
+          break;
+        }
+        case ABIReferenceType.asset: {
+          const uint64Type = new ABIUintType(64);
+          const refAssetID = uint64Type.decode(uint64Type.encode(refValue));
+          if (refAssetID > Number.MAX_SAFE_INTEGER) {
```

Comment:
separate minor question: what is the expected range of foreign asset/app id, do we have a expected upper/lower bound for that?

Reply:
Technically the range is 0 to the max uint64 value, but in practice actual asset & app IDs will be relatively "low" since they are always equal to the transaction sequence number.

And unfortunately the JS SDK uses a `number`, not a `bigint`, to represent IDs everywhere, so it has an implicit maximum value of [`Number.MAX_SAFE_INTEGER`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER). In v2 of the SDK we should change this to a `bigint` to be safe & more consistent with our other values.